File: src/com/crdroid/settings/fragments/lockscreen/UdfpsIconPicker.java
/*
 * Copyright (C) 2022 crDroid Android Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.crdroid.settings.fragments.lockscreen;

import android.content.ContentResolver;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.graphics.drawable.AnimationDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.net.Uri;
import android.provider.SearchIndexableResource;
import android.provider.Settings;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;
import androidx.preference.Preference.OnPreferenceChangeListener;
import androidx.preference.Preference;
import androidx.preference.PreferenceScreen;
import androidx.preference.PreferenceViewHolder;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import com.android.internal.logging.nano.MetricsProto.MetricsEvent;
import com.android.settings.R;
import com.android.settings.SettingsPreferenceFragment;
import com.android.settings.search.BaseSearchIndexProvider;
import com.android.settingslib.search.Indexable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.json.JSONException;
import org.json.JSONObject;

public class UdfpsIconPicker extends SettingsPreferenceFragment {

    private RecyclerView mRecyclerView;

    private Resources udfpsRes;

    private String mPkg = "com.crdroid.udfps.resources";

    private String[] mIcons;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getActivity().setTitle(R.string.udfps_icon_picker_title);

        loadResources();
    }

    private void loadResources() {
        try {
            PackageManager pm = getActivity().getPackageManager();
            udfpsRes = pm.getResourcesForApplication(mPkg);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }

        mIcons = udfpsRes.getStringArray(udfpsRes.getIdentifier("udfps_icons",
                "array", mPkg));
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(
                R.layout.item_view, container, false);

        mRecyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getActivity(), 3);
        mRecyclerView.setLayoutManager(gridLayoutManager);
        UdfpsIconAdapter mUdfpsIconAdapter = new UdfpsIconAdapter(getActivity());
        mRecyclerView.setAdapter(mUdfpsIconAdapter);

        return view;
    }

    @Override
    public int getMetricsCategory() {
        return MetricsEvent.CRDROID_SETTINGS;
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    public class UdfpsIconAdapter extends RecyclerView.Adapter<UdfpsIconAdapter.UdfpsIconViewHolder> {
        Context context;
        String mSelectedIcon;
        String mAppliedIcon;

        public UdfpsIconAdapter(Context context) {
            this.context = context;
        }

        @Override
        public UdfpsIconViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_option, parent, false);
            UdfpsIconViewHolder vh = new UdfpsIconViewHolder(v);
            return vh;
        }

        @Override
        public void onBindViewHolder(UdfpsIconViewHolder holder, final int position) {
            String iconRes = mIcons[position];

            Glide.with(holder.image.getContext())
                    .load("")
                    .placeholder(getDrawable(holder.image.getContext(), mIcons[position]))
                    .into(holder.image);

            holder.image.setPadding(20,20,20,20);

            holder.name.setVisibility(View.GONE);

            if (position == Settings.System.getInt(context.getContentResolver(),
                Settings.System.UDFPS_ICON, 0)) {
                mAppliedIcon = iconRes;
                if (mSelectedIcon == null) {
                    mSelectedIcon = iconRes;
                }
            }
            holder.itemView.setActivated(iconRes == mSelectedIcon);
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    updateActivatedStatus(mSelectedIcon, false);
                    updateActivatedStatus(iconRes, true);
                    mSelectedIcon = iconRes;
                    Settings.System.putInt(getActivity().getContentResolver(),
                            Settings.System.UDFPS_ICON, position);
                }
            });
        }

        @Override
        public int getItemCount() {
            return mIcons.length;
        }

        public class UdfpsIconViewHolder extends RecyclerView.ViewHolder {
            TextView name;
            ImageView image;
            public UdfpsIconViewHolder(View itemView) {
                super(itemView);
                name = (TextView) itemView.findViewById(R.id.option_label);
                image = (ImageView) itemView.findViewById(R.id.option_thumbnail);
            }
        }

        private void updateActivatedStatus(String icon, boolean isActivated) {
            int index = Arrays.asList(mIcons).indexOf(icon);
            if (index < 0) {
                return;
            }
            RecyclerView.ViewHolder holder = mRecyclerView.findViewHolderForAdapterPosition(index);
            if (holder != null && holder.itemView != null) {
                holder.itemView.setActivated(isActivated);
            }
        }
    }

    public Drawable getDrawable(Context context, String drawableName) {
        try {
            PackageManager pm = context.getPackageManager();
            Resources res = pm.getResourcesForApplication(mPkg);
            Context ctx = context.createPackageContext(
                    mPkg, Context.CONTEXT_IGNORE_SECURITY);
            return ctx.getDrawable(res.getIdentifier(drawableName, "drawable", mPkg));
        }
        catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: src/com/crdroid/settings/fragments/lockscreen/UdfpsIconPicker.kt
/*
 * SPDX-FileCopyrightText: crDroid Android Project
 * SPDX-License-Identifier: GPL-3.0
 */

package com.crdroid.settings.fragments.lockscreen

import android.content.ContentResolver
import android.content.Context
import android.content.pm.PackageManager
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.UserHandle
import android.provider.Settings
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import androidx.annotation.DrawableRes
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

import com.android.internal.logging.nano.MetricsProto.MetricsEvent
import com.android.settings.R
import com.android.settings.SettingsPreferenceFragment

class UdfpsIconPicker : SettingsPreferenceFragment() {

    private var udfpsPkg = "com.crdroid.udfps.icons"

    private var udfpsCtx: Context? = null
    private var udfpsRes: Resources? = null

    private val items = ArrayList<IconItem>(0)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requireActivity().setTitle(R.string.udfps_icon_picker_title)
        loadResources()
    }

    private fun loadResources() {
        items.clear()
        val host = activity ?: return

        try {
            val pkgCtx = host.createPackageContext(udfpsPkg, Context.CONTEXT_IGNORE_SECURITY)
            val res = pkgCtx.resources

            val arrayId = res.getIdentifier("udfps_icons", "array", udfpsPkg)
            if (arrayId == 0) {
                udfpsCtx = null
                udfpsRes = null
                return
            }

            val iconNames = res.getStringArray(arrayId)
            udfpsCtx = pkgCtx
            udfpsRes = res

            iconNames.forEachIndexed { index, name ->
                val id = res.getIdentifier(name, "drawable", udfpsPkg)
                items.add(
                    IconItem(
                        index = index,
                        name = name,
                        drawableId = id
                    )
                )
            }
        } catch (_: PackageManager.NameNotFoundException) {
            udfpsCtx = null
            udfpsRes = null
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.item_view, container, false)
        val rv = view.findViewById<RecyclerView>(R.id.recycler_view)

        val spanCount = if (resources.configuration.orientation ==
            android.content.res.Configuration.ORIENTATION_LANDSCAPE
        ) 4 else 3

        rv.layoutManager = GridLayoutManager(requireContext(), spanCount)

        val adapter = UdfpsIconAdapter(
            contentResolver = requireContext().contentResolver,
            iconLoader = { item -> getIconDrawable(item.drawableId) }
        )
        rv.adapter = adapter
        adapter.submitList(items.toList())

        return view
    }

    private fun getIconDrawable(@DrawableRes id: Int): Drawable? {
        val c = udfpsCtx ?: return null
        if (id == 0) return null
        return try {
            c.getDrawable(id)
        } catch (_: Resources.NotFoundException) {
            null
        }
    }

    override fun getMetricsCategory(): Int = MetricsEvent.CRDROID_SETTINGS

    companion object {
        @JvmStatic
        fun reset(context: Context) {
            val resolver: ContentResolver = context.contentResolver
            Settings.System.putIntForUser(
                resolver,
                Settings.System.UDFPS_ICON,
                0,
                UserHandle.USER_CURRENT
            )
        }
    }

    private data class IconItem(
        val index: Int,
        val name: String,
        @DrawableRes val drawableId: Int
    )

    private class UdfpsIconAdapter(
        private val contentResolver: ContentResolver,
        private val iconLoader: (IconItem) -> Drawable?
    ) : ListAdapter<IconItem, UdfpsIconAdapter.VH>(DIFF) {

        private var selectedPos: Int = -1

        init {
            setHasStableIds(true)
        }

        override fun getItemId(position: Int): Long = getItem(position).index.toLong()

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
            val v = LayoutInflater.from(parent.context).inflate(R.layout.item_option, parent, false)
            return VH(v)
        }

        override fun onBindViewHolder(holder: VH, position: Int) {
            val item = getItem(position)

            if (selectedPos == -1) {
                selectedPos = Settings.System.getIntForUser(
                    contentResolver,
                    Settings.System.UDFPS_ICON,
                    0,
                    UserHandle.USER_CURRENT
                )
            }

            holder.image.setImageDrawable(iconLoader(item))
            holder.image.setPadding(20, 20, 20, 20)
            holder.name.visibility = View.GONE

            holder.itemView.isActivated = (position == selectedPos)

            holder.itemView.setOnClickListener {
                val old = selectedPos
                selectedPos = position
                if (old != -1) notifyItemChanged(old)
                notifyItemChanged(selectedPos)

                Settings.System.putIntForUser(
                    contentResolver,
                    Settings.System.UDFPS_ICON,
                    position,
                    UserHandle.USER_CURRENT
                )
            }
        }

        class VH(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val name: TextView = itemView.findViewById(R.id.option_label)
            val image: ImageView = itemView.findViewById(R.id.option_thumbnail)
        }

        private companion object {
            val DIFF = object : DiffUtil.ItemCallback<IconItem>() {
                override fun areItemsTheSame(oldItem: IconItem, newItem: IconItem): Boolean {
                    return oldItem.index == newItem.index
                }

                override fun areContentsTheSame(oldItem: IconItem, newItem: IconItem): Boolean {
                    return oldItem == newItem
                }
            }
        }
    }
}
